# Resolução de Modelos de Circuitos Elétricos em Sistemas Lineares pelo Sympy

O circuito representado pela figura:

![](https://diegoascanio.github.io/ace-aula-03/img/circuito_final.png)

Pode ser modelado pelas leis de kirchoff da corrente e da tensão para o seguinte sistema de equações lineares:

$$
\begin{cases}
    &1 i_{0} - 1 i_{1} - 1 i_{2} + 0 i_{3} + 0 i_{4} + 0 i_{5} &= 0\\
    &0 i_{0} + 0 i_{1} + 1 i_{2} - 1 i_{3} - 1 i_{4} + 0 i_{5} &= 0\\
    &0 i_{0} - 1 i_{1} + 0 i_{2} - 1 i_{3} + 0 i_{4} + 1 i_{5} &= 0\\
    &0 i_{0} + 0 i_{1} - R i_{2} + 0 i_{3} - 30 i_{4} + 0 i_{5} &= -125\\
    &0 i_{0} - 15 i_{1} + R i_{2} + 5 i_{3} + 0 i_{4} + 0 i_{5} &= 0\\
    &0 i_{0} + 0 i_{1} + 0 i_{2} - 5 i_{3} + 30 i_{4} - 16 i_{5} &= 0\\
\end{cases}
$$

A matriz extendida composta pela matriz de coeficientes das variáveis dos sistemas e de seus termos independentes é dada por:

$$
\begin{bmatrix}
    1 & -1 & -1 & 0 & 0 & 0 & 0\\
    0 & 0 & 1 & -1 & -1 & 0 & 0\\
    0 & -1 & 0 & -1 & 0 & 1 & 0\\
    0 & 0 & -R & 0 & -30 & 0 & -125\\
    0 & -15 & R & 5 & 0 & 0 & 0\\
    0 & 0 & 0 & -5 & 30 & -16 & 0\\
\end{bmatrix}
$$

O SymPy é uma biblioteca em python que permite realizar álgebra simbólica, ou seja, considerar as variáveis apenas pelos seus símbolos, sem a necessidade de atribuir valores.

Para usarmos a biblioteca sympy no jupyter lite é necessário instalá-la primeiro:


In [1]:
%pip install sympy

Uma vez que a biblioteca foi instalada, devemos importar os recursos que oferece:

In [2]:
from sympy import *

Definir nosso símbolo do resistor desconhecido $R$

In [3]:
R = Symbol('R')

E definir nossa matriz extendida:

In [4]:
M = Matrix([
    [1, -1, -1, 0, 0, 0, 0],
    [0, 0, 1, -1, -1, 0, 0],
    [0, -1, 0, -1, 0, 1, 0],
    [0, 0, -R, 0, -30, 0, -125],
    [0, -15, R, 5, 0, 0, 0],
    [0, 0, 0, -5, 30, -16, 0]
])

Uma vez definida nossa matriz, conseguimos obter todas as correntes $i_{0} \cdots i_{6}$ em função de $R$ através da forma escada reduzida por linhas de $M$, obtida pelo método rref:

In [5]:
M_escada_linha_reduzida, indices_correntes = M.rref()
M_escada_linha_reduzida

Matrix([
[1, 0, 0, 0, 0, 0, (255*R + 4975)/(53*R + 474)],
[0, 1, 0, 0, 0, 0,  (255*R + 750)/(53*R + 474)],
[0, 0, 1, 0, 0, 0,           4225/(53*R + 474)],
[0, 0, 0, 1, 0, 0,  (2250 - 80*R)/(53*R + 474)],
[0, 0, 0, 0, 1, 0,  (80*R + 1975)/(53*R + 474)],
[0, 0, 0, 0, 0, 1, (175*R + 3000)/(53*R + 474)]])

Agora, temos todas as correntes escritas em função do resistor desconhecido $R$, observe que além de retornar a matriz escada linha reduzida, rref retorna também os indices das variáveis que buscamos - no nosso caso a corrente elétrica de cada um dos elementos do circuito. No começo do exemplo, foi nos informado que a corrente que passa pelo resistor de $9\Omega$ é de $3A$. Essa corrente é $i_1$.

A equação de $i_1$ em função de $R$ está na ultima coluna da segunda linha da matriz escada linha reduzida:

In [8]:
i_1_R = M_escada_linha_reduzida[1,-1]
i_1_R

(255*R + 750)/(53*R + 474)

Sympy tem o método solve, que obtém a raiz (o valor da variável simbólica quando a equação é igual a zero) de uma equação informada.

Sabemos que $i_1 - 3A = \frac{255 R + 750}{53 R + 474} - 3A = 0$

Se resolvermos esta equação, logo, encontraremos o valor de R.

In [9]:
solve(i_1_R - 3)

[7]

Como $R = 7 \Omega$, ao substituir este valor nas equações das outras correntes, resolvemos o circuito por completo:

In [14]:
# defino um vetor de correntes
I = {}
# R vale 7 Ohms
r = 7
# resolvo as correntes i0 até i5 em função de R
for corrente in indices_correntes:
    funcao_corrente_em_funcao_de_r = M_escada_linha_reduzida[corrente, -1]
    # o subs, substitui o Valor R desconhecido pelo r resolvido
    # que vale 7 ohms
    corrente_em_funcao_de_r = funcao_corrente_em_funcao_de_r.subs(
        R, r
    )
    I['i_{}'.format(corrente)] = '{}A'.format(corrente_em_funcao_de_r)
# mostro as correntes do circuito
I

{'i_0': '8A', 'i_1': '3A', 'i_2': '5A', 'i_3': '2A', 'i_4': '3A', 'i_5': '5A'}

Assim, nosso circuito está resolvido!